clients/pubsub-common/src/test/java/org/eclipse/hono/client/pubsub/AbstractPubSubBasedMessageSenderTest.java

```diff
@@ -0,0 +1,222 @@
+
+        final PubSubPublisherClient client = mock(PubSubPublisherClient.class);
+        when(client.publish(Mockito.any())).thenReturn(
+                Future.failedFuture(new ClientErrorException(HttpURLConnection.HTTP_CONFLICT, "client is null")));
+        when(factory.getOrCreatePublisher(TOPIC, PROJECT_ID, TENANT_ID)).thenReturn(client);
+
+        try (MockedStatic<PubsubMessage> pubSubMessage = Mockito.mockStatic(PubsubMessage.class)) {
+            final PubsubMessage.Builder builder = mock(PubsubMessage.Builder.class);
+            when(builder.putAllAttributes(Mockito.any())).thenReturn(builder);
+            when(builder.setOrderingKey(DEVICE_ID)).thenReturn(builder);
+            when(builder.setData(Mockito.any())).thenReturn(builder);
+
+            final PubsubMessage message = mock(PubsubMessage.class);
+            when(builder.build()).thenReturn(message);
+
+            pubSubMessage.when(PubsubMessage::newBuilder).thenReturn(builder);
+
+            assertThrows(NullPointerException.class,
+                    () -> sender.sendAndWaitForOutcome(TOPIC, TENANT_ID, DEVICE_ID, null, Map.of(), NoopSpan.INSTANCE));
+
+            assertThrows(ServerErrorException.class,
+                    () -> sender.sendAndWaitForOutcome(TOPIC, TENANT_ID, DEVICE_ID, Buffer.buffer("test payload"),
+                            Map.of(), NoopSpan.INSTANCE));
+        }
+
+    }
+
+    /**
+     * Verifies that the send method succeeded.
+     */
+    @Test
+    public void testSendSucceeded() {
+        final Map<String, Object> properties = getProperties();
+        final Map<String, String> attributes = getAttributes();
+
+        when(lifecycleStatus.isStarted()).thenReturn(true);
+
+        final PubSubPublisherClient client = mock(PubSubPublisherClient.class);
+        when(client.publish(Mockito.any())).thenReturn(Future.succeededFuture());
+        when(factory.getOrCreatePublisher(TOPIC, PROJECT_ID, TENANT_ID)).thenReturn(client);
+
+        try (MockedStatic<PubsubMessage> pubSubMessage = Mockito.mockStatic(PubsubMessage.class)) {
+            final PubsubMessage.Builder builder = mock(PubsubMessage.Builder.class);
+            when(builder.putAllAttributes(attributes)).thenReturn(builder);
+            when(builder.setOrderingKey(DEVICE_ID)).thenReturn(builder);
+            final byte[] b = new byte[22];
+            new Random().nextBytes(b);
+            final ByteString bs = ByteString.copyFrom(b);
+            when(builder.setData(bs)).thenReturn(builder);
+
+            final PubsubMessage message = mock(PubsubMessage.class);
+            when(builder.build()).thenReturn(message);
+
+            pubSubMessage.when(PubsubMessage::newBuilder).thenReturn(builder);
+
+            final var result = sender.sendAndWaitForOutcome(TOPIC, TENANT_ID, DEVICE_ID,
+                    Buffer.buffer(b),
+                    properties,
+                    NoopSpan.INSTANCE);
+            assertThat(result.succeeded()).isTrue();
```

Comment:
FMPOV we should add some assertions regarding the message that has been published, i.e. the message payload and its properties. In the end that is the functionality of the sender, creating a corresponding Pub/Sub message from the generic inputs ...

Reply:
We think this doesn't make sense since the message that has been published is mocked inside this test. In our testcase we make sure we use the right attributes (that match to the properties) in the mock. Otherwise it won't work.
If you want to have the attributes tested we can test the private method `encodePropertiesAsPubSubAttributes(final Map<String, Object> properties, final Span span)` ?

Reply:
Yes, updated the test